File: problems/3495_minimum_operations_to_make_array_elements_zero/3495_minimum_operations_to_make_array_elements_zero.py
class Solution:
    def minOperations(self, queries: List[List[int]]) -> int:
        return sum((self.ops(r) - self.ops(l - 1) + 1) // 2 for l, r in queries)
    
    def ops(self, n: int) -> int:
        res, steps, base = 0, 0, 1
        while base <= n:
            steps += 1
            end = min (n, base * 4 - 1)
            res += (end - base + 1) * steps
            base *= 4
        return res

File: problems/3495_minimum_operations_to_make_array_elements_zero/3495_minimum_operations_to_make_array_elements_zero.java
class Solution {
    public long minOperations(int[][] queries) {
        long ans = 0;
        for (int q[] : queries) {
            int l = q[0], r = q[1];
            ans += (getOps(r) - getOps(l - 1) + 1) / 2;
        }
        return ans;
    }

    private long getOps(int n) {
        long res = 0;
        int steps = 0;
        for (long base = 1; base <= n; base *= 4) {
            steps++;
            long end = Math.min(n, base * 4 - 1);
            res += (end - base + 1) * steps;
        }
        return res;
    }
}